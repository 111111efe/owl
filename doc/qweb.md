# ðŸ¦‰ QWeb ðŸ¦‰

## Content

- [Overview](#overview)
- [QWeb Specification](#qweb-specification)
    - [Static html nodes](#static-html-nodes)
    - [`t-esc` directive](#t-esc-directive)
    - [`t-raw` directive](#t-raw-directive)
    - [`t-set` directive](#t-set-directive)
    - [`t-if` directive](#t-if-directive)
    - [Expression evaluation](#expression-evaluation)
    - [`t-att` directive (dynamic attributes)](#t-att-directive-dynamic-attributes)
- [JS/OWL Specific Extensions](#jsowl-specific-extensions)
    - [`t-on` directive](#t-on-directive)
    - [Component: `t-widget`, `t-props`](#component-t-widget-t-props)
    - [`t-ref` directive](#t-ref-directive)
    - [`t-key` directive](#t-key-directive)
    - [Debugging (`t-debug` and `t-log`)](#debugging-t-debug-and-t-log)
    - [White spaces](#white-spaces)
    - [Root nodes](#root-nodes)

## Overview

[QWeb](https://www.odoo.com/documentation/12.0/reference/qweb.html) is the primary templating engine used by Odoo. It is based on the XML format, and used
mostly to generate html. In OWL, QWeb templates are compiled into functions that
generate a virtual dom representation of the html.

Template directives are specified as XML attributes prefixed with `t-`, for instance `t-if` for conditionals, with elements and other attributes being rendered directly.

To avoid element rendering, a placeholder element `<t>` is also available, which executes its directive but doesnâ€™t generate any output in and of itself.

The QWeb implementation in the OWL project is slightly different. It compiles
templates into functions that output a virtual DOM instead of a string. This is
necessary for the component system. In addition, it has a few extra directives
(see [OWL Specific Extensions](#owlspecificextensions))


## QWeb Specification


### Static html nodes

Normal, regular html nodes are rendered into themselves:

```xml
  <div>hello</div> <!â€“â€“ rendered as itself â€“â€“>
```

### `t-esc` directive

The `t-esc` directive is necessary whenever you want to add a dynamic text
expression in a template. The text is escaped to avoid security issues.

```xml
<p><t t-esc="value"/></p>
```

rendered with the value `value` set to `42` in the rendering context yields:

```html
<p>42</p>
```

### `t-raw` directive

The `t-raw` directive is almost the same as `t-esc`, but without the escaping.
This is mostly useful to inject a raw html string somewhere. Obviously, this
is unsafe to do in general, and should only be used for strings known to be safe.

```xml
<p><t t-esc="value"/></p>
```

rendered with the value `value` set to `<span>foo</span>` in the rendering context yields:

```html
<p><span>foo</span></p>
```

### `t-set` directive

QWeb allows creating variables from within the template, to memoize a computation (to use it multiple times), give a piece of data a clearer name, ...

This is done via the `t-set` directive, which takes the name of the variable to create. The value to set can be provided in two ways:

1. a `t-value` attribute containing an expression, and the result of its
   evaluation will be set:

   ```xml
   <t t-set="foo" t-value="2 + 1"/>
   <t t-esc="foo"/>
   ```

   will print `3`. Note that the evaluation is done at rendering time, not at
   compilte time.

2. if there is no `t-value` attribute, the nodeâ€™s body is saved and its value is
   set as the variableâ€™s value:

   ```xml
   <t t-set="foo">
       <li>ok</li>
   </t>
   <t t-esc="foo"/>
   ```

   will generate `&lt;li&gt;ok&lt;/li&gt;` (the content is escaped as we used the `t-esc` directive)

The `t-set` directive acts like a regular variable in most programming language.
It is lexically scoped (inner nodes are sub scopes), can be shadowed, ...

### `t-if` directive

The `t-if` directive is useful to conditionally render something. It evaluates
the expression given as attribute value, and then acts accordingly.

```xml
<div>
    <t t-if="condition">
        <p>ok</p>
    </t>
</div>
```

The element is rendered if the condition (evaluated with the current rendering
context) is true:

```xml
<div>
    <p>ok</p>
</div>
```

but if the condition is false it is removed from the result:

```xml
<div>
</div>
```

The conditional rendering applies to the bearer of the directive, which does not
have to be `<t>`:

```xml
<div>
    <p t-if="condition">ok</p>
</div>
```

will give the same results as the previous example.

Extra conditional branching directives `t-elif` and `t-else` are also available:

```xml
<div>
    <p t-if="user.birthday == today()">Happy bithday!</p>
    <p t-elif="user.login == 'root'">Welcome master!</p>
    <p t-else="">Welcome!</p>
</div>
```

### Expression evaluation

It is useful to explain the various rules that applies on QWeb expressions. These
expressions are strings that will be converted to a javascript expression at
compile time.

1. it should be a simple expression which returns a value. It cannot be a statement.

   ```xml
   <div><p t-if="1 + 2 === 3">ok</p></div>
   ```

   is valid, but the following is not valid:

   ```xml
   <div><p t-if="console.log(1)">NOT valid</p></div>
   ```

2. it can use anything in the rendering context:

   ```xml
   <p t-if="user.birthday == today()">Happy bithday!</p>
   ```

   is valid, and will read the `user` object from the context, and call the
   `today` function.

3. it can use a few special operators to avoid using symbols such as `<`, `>`,
   `&` or `|`. This is useful to make sure that we still write valid XML.

   | Word  | will be replaced by |
   | ----- | ------------------- |
   | `and` | `&&`                |
   | `or`  | `\|\|`                |
   | `gt`  | `>`                 |
   | `gte` | `>=`                |
   | `lt`  | `<`                 |
   | `lte` | `<=`                |

   So, one can write this:

   ```xml
   <div><p t-if="10 + 2 gt 5">ok</p></div>
   ```

### `t-att` directive (dynamic attributes)

One can use the `t-att-` directive to add dynamic attributes. Its main use is to
evaluate an expression (at rendering time) and bind an attribute to its result:

For example, if we have `id` set to 32 in the rendering context,

```xml
<div t-att-data-action-id="id"/>  <!-- result: <div data-action-id="32"></div> -->
```

If an expression evaluates to a falsy value, it will not be set at all:

```xml
<div t-att-foo="false"/>  <!-- result: <div></div> -->
```

## JS/OWL Specific Extensions

### `t-on` directive

### Component: `t-widget`, `t-props`

### `t-ref` directive

### `t-key` directive

Even though Owl tries to be as declarative as possible, some DOM state is still
locked inside the DOM: for example, the scrolling state, the current user selection,
the focused element or the state of an input. This is why we use a virtual dom
algorithm to keep the actual DOM node as much as possible.  However, this is
sometimes not enough, and we need to help Owl decide if an element is actually
the same, or is different. The `t-key` directive is used to give an identity to an element.

There are three main use cases:

- *elements in a list*:
  ```xml
    <span t-foreach="todos" t-as="todo" t-key="todo.id">
        <t t-esc="todo.text"/>
    </span>
  ```

- *`t-if`/`t-else`*

- *animations*: give a different identity to a component.  Ex: thread id with
animations on add/remove message.

### Debugging (`t-debug` and `t-log`)

The javascript QWeb implementation provides two useful debugging directives:

`t-debug` adds a debugger statement during template rendering:

```xml
<t t-if="a_test">
    <t t-debug="">
</t>
````

will stop execution if the browser dev tools are open.

`t-log` takes an expression parameter, evaluates the expression during rendering and logs its result with console.log:

```xml
<t t-set="foo" t-value="42"/>
<t t-log="foo"/>
```

will print 42 to the console


### White spaces

White spaces in a templates are handled in a special way:

- consecutive whitespaces are always condensed to a single whitespace
- if a whitespace-only text node contains a linebreak, it is ignored
- the previous rules do not apply if we are in a `<pre>` tag

### Root nodes

For many reasons, Owl QWeb templates should have a single root node.  More
precisely, the result of a template rendering should have a single root node:

```xml
<!â€“â€“ not ok: two root nodes â€“â€“>
<t>
    <div>foo</div>
    <div>bar</div>
</t>

<!â€“â€“ ok: result has one single root node â€“â€“>
<t>
    <div t-if="someCondition">foo</div>
    <span t-else="1">bar</span>
</t>
```

Extra root nodes will actually be ignored (even though they will be rendered
in memory).

Note: this does not apply to subtemplates (see the `t-call` directive).  In that
case, they will be inlined in the main template, and can actually have many
root nodes.