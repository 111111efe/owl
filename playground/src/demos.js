export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
