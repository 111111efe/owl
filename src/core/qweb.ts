import { escape } from "./utils";
import directives from "./qweb_directives";
import { Directive } from "./qweb_directives";

type RawTemplate = string;
type Template = (context: any) => DocumentFragment;

// Evaluation Context
export type EvalContext = { [key: string]: any };

const RESERVED_WORDS = "true,false,NaN,null,undefined,debugger,console,window,in,instanceof,new,function,return,this,typeof,eval,void,Math,RegExp,Array,Object,Date".split(
  ","
);

// Compilation Context
export class Context {
  nextID: number = 1;
  code: string[] = [];
  variables: { [key: string]: any } = {};
  escaping: boolean = false;
  parentNode: string | undefined;
  indentLevel: number = 0;
  rootContext: Context;
  caller: Element | undefined;
  fragmentID: string;

  constructor() {
    this.rootContext = this;
    this.fragmentID = this.generateID();
  }

  generateID(): string {
    const id = `_${this.rootContext.nextID}`;
    this.rootContext.nextID++;
    return id;
  }

  withParent(node: string): Context {
    const newContext = Object.create(this);
    newContext.parentNode = node;
    return newContext;
  }

  withVariables(variables: { [key: string]: any }) {
    const newContext = Object.create(this);
    newContext.variables = Object.create(variables);
    return newContext;
  }

  withCaller(node: Element): Context {
    const newContext = Object.create(this);
    newContext.caller = node;
    return newContext;
  }

  withEscaping(): Context {
    const newContext = Object.create(this);
    newContext.escaping = true;
    return newContext;
  }

  indent() {
    this.indentLevel++;
  }

  dedent() {
    this.indentLevel--;
  }

  addNode(nodeID: string) {
    if (this.parentNode) {
      this.addLine(`${this.parentNode}.appendChild(${nodeID})`);
    } else {
      this.addLine(`${this.fragmentID}.appendChild(${nodeID})`);
    }
  }
  addLine(line: string) {
    const prefix = new Array(this.indentLevel).join("\t");
    const lastChar = line[line.length - 1];
    const suffix = lastChar !== "}" && lastChar !== "{" ? ";" : "";
    this.code.push(prefix + line + suffix);
  }
}

/**
 * Template rendering engine
 */
export default class QWeb {
  rawTemplates: { [name: string]: RawTemplate } = {};
  nodeTemplates: { [name: string]: Document } = {};
  templates: { [name: string]: Template } = {};
  escape: ((str: string) => string) = escape;
  exprCache: { [key: string]: string } = {};
  directives: Directive[] = [];

  constructor() {
    directives.forEach(d => this.addDirective(d));
  }

  addDirective(dir: Directive) {
    this.directives.push(dir);
    this.directives.sort((d1, d2) => d1.priority - d2.priority);
  }
  /**
   * Add a template to the internal template map.  Note that it is not
   * immediately compiled.
   */
  addTemplate(name: string, template: RawTemplate) {
    if (name in this.rawTemplates) {
      return;
    }
    this.rawTemplates[name] = template;
    const parser = new DOMParser();
    const doc = parser.parseFromString(template, "text/xml");
    if (!doc.firstChild) {
      throw new Error("Invalid template (should not be empty)");
    }
    if (doc.firstChild.nodeName === "parsererror") {
      throw new Error("Invalid XML in template");
    }
    let tbranch = doc.querySelectorAll("[t-elif], [t-else]");
    for (let i = 0, ilen = tbranch.length; i < ilen; i++) {
      let node = tbranch[i];
      let prevElem = node.previousElementSibling!;
      let pattr = function(name) {
        return prevElem.getAttribute(name);
      };
      let nattr = function(name) {
        return +!!node.getAttribute(name);
      };
      if (prevElem && (pattr("t-if") || pattr("t-elif"))) {
        if (pattr("t-foreach")) {
          throw new Error(
            "t-if cannot stay at the same level as t-foreach when using t-elif or t-else"
          );
        }
        if (
          ["t-if", "t-elif", "t-else"].map(nattr).reduce(function(a, b) {
            return a + b;
          }) > 1
        ) {
          throw new Error(
            "Only one conditional branching directive is allowed per node"
          );
        }
        // All text nodes between branch nodes are removed
        let textNode;
        while ((textNode = node.previousSibling) !== prevElem) {
          if (textNode.nodeValue.trim().length) {
            throw new Error("text is not allowed between branching directives");
          }
          textNode.remove();
        }
      } else {
        throw new Error(
          "t-elif and t-else directives must be preceded by a t-if or t-elif directive"
        );
      }
    }

    this.nodeTemplates[name] = doc;
  }

  /**
   * Render a template
   *
   * @param {string} name the template should already have been added
   */
  render(name: string, context: any = {}): DocumentFragment {
    const template = this.templates[name] || this._compile(name);
    return template(context);
  }

  renderToString(name: string, context: EvalContext = {}): string {
    const node = this.render(name, context);
    return this._renderNodeToString(node);
  }

  _renderNodeToString(node: Node): string {
    switch (node.nodeType) {
      case 3: // text node
        return node.textContent!;
      case 11: // document.fragment
        const children = Array.from((<DocumentFragment>node).childNodes);
        return children.map(this._renderNodeToString).join("");
      default:
        return (<HTMLElement>node).outerHTML;
    }
  }

  _compile(name: string): Template {
    if (name in this.templates) {
      return this.templates[name];
    }
    const doc = this.nodeTemplates[name];

    let ctx = new Context();

    const mainNode = doc.firstChild!;
    ctx.addLine(`let ${ctx.fragmentID} = document.createDocumentFragment()`);
    this._compileNode(mainNode, ctx);

    ctx.addLine(`return ${ctx.fragmentID}`);
    const functionCode = ctx.code.join("\n");
    if ((<Element>mainNode).attributes.hasOwnProperty("t-debug")) {
      console.log(
        `Template: ${this.rawTemplates[name]}\nCompiled code:\n` + functionCode
      );
    }
    const template: Template = (new Function(
      "context",
      functionCode
    ) as Template).bind(this);
    this.templates[name] = template;
    return template;
  }

  /**
   * Generate code from an xml node
   *
   */
  _compileNode(node: ChildNode, ctx: Context) {
    if (!(node instanceof Element)) {
      // this is a text node, there are no directive to apply
      let text = node.textContent!;
      let nodeID = ctx.generateID();
      ctx.addLine(`let ${nodeID} = document.createTextNode(\`${text}\`)`);
      ctx.addNode(nodeID);
      return;
    }

    const attributes = (<Element>node).attributes;

    const validDirectives: {
      directive: Directive;
      value: string;
      fullName: string;
    }[] = [];

    for (let directive of this.directives) {
      // const value = attributes[i].textContent!;
      let fullName;
      let value;
      for (let i = 0; i < attributes.length; i++) {
        const name = attributes[i].name;
        if (
          name === "t-" + directive.name ||
          name.startsWith("t-" + directive.name + "-")
        ) {
          fullName = name;
          value = attributes[i].textContent;
        }
      }
      if (fullName) {
        validDirectives.push({ directive, value, fullName });
      }
    }

    for (let { directive, value, fullName } of validDirectives) {
      if (directive.atNodeEncounter) {
        const isDone = directive.atNodeEncounter({
          node,
          qweb: this,
          ctx,
          fullName,
          value
        });
        if (isDone) {
          return;
        }
      }
    }

    if (node.nodeName !== "t") {
      let nodeID = this._compileGenericNode(node, ctx);
      ctx = ctx.withParent(nodeID);

      for (let { directive, value, fullName } of validDirectives) {
        if (directive.atNodeCreation) {
          directive.atNodeCreation({
            node,
            qweb: this,
            ctx,
            fullName,
            value,
            nodeID
          });
        }
      }
    }

    this._compileChildren(node, ctx);

    for (let { directive, value, fullName } of validDirectives) {
      if (directive.finalize) {
        directive.finalize({ node, qweb: this, ctx, fullName, value });
      }
    }
  }

  _getValue(val: any, ctx: Context): any {
    if (val in ctx.variables) {
      return this._getValue(ctx.variables[val], ctx);
    }
    return val;
  }
  _compileChildren(node: ChildNode, ctx: Context) {
    if (node.childNodes.length > 0) {
      for (let child of Array.from(node.childNodes)) {
        this._compileNode(child, ctx);
      }
    }
  }
  _compileGenericNode(node: ChildNode, ctx: Context): string {
    let nodeID: string | undefined;
    switch (node.nodeType) {
      case 1: // generic tag;
        nodeID = ctx.generateID();
        ctx.addLine(
          `let ${nodeID} = document.createElement('${node.nodeName}')`
        );

        const attributes = (<Element>node).attributes;
        for (let i = 0; i < attributes.length; i++) {
          const name = attributes[i].name;
          const value = attributes[i].textContent!;
          if (!name.startsWith("t-")) {
            ctx.addLine(
              `${nodeID}.setAttribute('${name}', '${escape(value)}')`
            );
          }
          if (name.startsWith("t-att-")) {
            const attName = name.slice(6);
            const formattedValue = this._formatExpression(value!);
            const attID = ctx.generateID();
            ctx.addLine(`let ${attID} = ${formattedValue}`);
            ctx.addLine(
              `if (${attID}) {${nodeID}.setAttribute('${attName}', ${attID})}`
            );
          }
          if (name.startsWith("t-attf-")) {
            const exprName = name.slice(7);
            const formattedExpr = value!.replace(
              /\{\{.*?\}\}/g,
              s => "${" + this._formatExpression(s.slice(2, -2)) + "}"
            );
            ctx.addLine(
              `${nodeID}.setAttribute('${exprName}', \`${formattedExpr}\`)`
            );
          }
          if (name === "t-att") {
            const id = ctx.generateID();
            ctx.addLine(`let ${id} = ${this._formatExpression(value!)}`);
            ctx.addLine(`if (${id} instanceof Array) {`);
            ctx.indent();
            ctx.addLine(`${nodeID}.setAttribute(${id}[0], ${id}[1])`);
            ctx.dedent();
            ctx.addLine(`} else {`);
            ctx.indent();
            ctx.addLine(`for (let key in ${id}) {`);
            ctx.indent();
            ctx.addLine(`${nodeID}.setAttribute(key, ${id}[key])`);
            ctx.dedent();
            ctx.addLine(`}`);
            ctx.dedent();
            ctx.addLine(`}`);
          }
        }
        break;
      default:
        throw new Error("unknown node type");
    }

    ctx.addNode(nodeID);
    return nodeID;
  }

  _formatExpression(e: string): string {
    if (e in this.exprCache) {
      return this.exprCache[e];
    }
    // Thanks CHM for this code...
    const chars = e.split("");
    let instring = "";
    let invar = "";
    let invarPos = 0;
    let r = "";
    chars.push(" ");
    for (var i = 0, ilen = chars.length; i < ilen; i++) {
      var c = chars[i];
      if (instring.length) {
        if (c === instring && chars[i - 1] !== "\\") {
          instring = "";
        }
      } else if (c === '"' || c === "'") {
        instring = c;
      } else if (c.match(/[a-zA-Z_\$]/) && !invar.length) {
        invar = c;
        invarPos = i;
        continue;
      } else if (c.match(/\W/) && invar.length) {
        // TODO: Should check for possible spaces before dot
        if (chars[invarPos - 1] !== "." && RESERVED_WORDS.indexOf(invar) < 0) {
          invar = "context['" + invar + "']";
        }
        r += invar;
        invar = "";
      } else if (invar.length) {
        invar += c;
        continue;
      }
      r += c;
    }
    const result = r.slice(0, -1);
    this.exprCache[e] = result;
    return result;
  }
}
