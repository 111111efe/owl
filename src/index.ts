import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
