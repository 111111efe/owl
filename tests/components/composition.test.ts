import { Component, mount, useState, xml } from "../../src/index";
import { qweb } from "../../src/qweb/qweb";
import { makeTestFixture, nextTick } from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe("component composition", () => {
  describe("basic properties", () => {
    test("a functional component inside another", async () => {
      const Child = {
        template: xml`<div>simple vnode</div>`,
      };
      const Parent = {
        template: xml`<span><Child/></span>`,
        setup() {
          return { Child };
        },
      };

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a functional component inside another, alternate definition", async () => {
      const Child = {
        template: xml`<div>simple vnode</div>`,
      };
      const Parent = {
        template: xml`<span><Child/></span>`,
        components: { Child },
      };

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("props are passed from parent to child", async () => {
      const Child = {
        template: xml`<div><t t-esc="props.value"/></div>`,
      };
      const Parent = {
        template: xml`<span><Child value="v"/></span>`,
        setup() {
          return { Child, v: 123 };
        },
      };

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>123</div></span>");
    });

    test("a class component inside a function component", async () => {
      class Child extends Component {
        static template = xml`<div>simple vnode</div>`;
      }

      const Parent = {
        template: xml`<span><Child/></span>`,
        setup() {
          return { Child };
        },
      };
      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a class component inside a function component, alternate definition", async () => {
      class Child extends Component {
        static template = xml`<div>simple vnode</div>`;
      }

      const Parent = {
        template: xml`<span><Child/></span>`,
        components: { Child },
      };
      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a function component inside a class component", async () => {
      const Child = {
        template: xml`<div>simple vnode</div>`,
      };

      class Parent extends Component {
        static template = xml`<span><Child/></span>`;
        Child = Child;
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a function component inside a class component, alternate definition", async () => {
      const Child = {
        template: xml`<div>simple vnode</div>`,
      };

      class Parent extends Component {
        static template = xml`<span><Child/></span>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a class component inside a class component", async () => {
      class Child extends Component {
        static template = xml`<div>simple vnode</div>`;
      }

      class Parent extends Component {
        static template = xml`<span><Child/></span>`;
        Child = Child;
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a class component inside a class component, alternate definition", async () => {
      class Child extends Component {
        static template = xml`<div>simple vnode</div>`;
      }

      class Parent extends Component {
        static template = xml`<span><Child/></span>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<span><div>simple vnode</div></span>");
    });

    test("a class component with state inside a class component", async () => {
      class Child extends Component {
        static template = xml`<button t-esc="state.value" t-on-click="state.value++" />`;
        state = useState({ value: 1 });
      }

      class Parent extends Component {
        static template = xml`<span><Child/></span>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      fixture.querySelector("button")!.click();
      await nextTick();
      expect(fixture.innerHTML).toBe("<span><button>2</button></span>");
    });

    test("a function component with state inside a function component", async () => {
      const Child = {
        template: xml`<button t-esc="state.value" t-on-click="state.value++" />`,
        setup() {
          return { state: useState({ value: 1 }) };
        },
      };

      const Parent = {
        template: xml`<span><Child/></span>`,
        components: { Child },
      };

      await mount(Parent, fixture);
      fixture.querySelector("button")!.click();
      await nextTick();
      expect(fixture.innerHTML).toBe("<span><button>2</button></span>");
    });
  });

  describe("various update scenarios", () => {
    test("reconciliation alg is not confused in some specific situation", async () => {
      // in this test, we set t-key to 4 because it was in conflict with the
      // template id corresponding to the first child.
      class Child extends Component {
        static template = xml`<span>child</span>`;
      }

      class Parent extends Component {
        static template = xml`
              <div>
                  <Child />
                  <Child t-key="4"/>
              </div>
            `;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<div><span>child</span><span>child</span></div>");
    });

    test("reconciliation alg works for t-foreach in t-foreach", async () => {
      // const warn = console.warn;
      // console.warn = () => {};
      class Child extends Component {
        static template = xml`<div><t t-esc="props.blip"/></div>`;
      }

      class Parent extends Component {
        static template = xml`
              <div>
                  <t t-foreach="state.s" t-as="section">
                      <t t-foreach="section.blips" t-as="blip">
                        <Child blip="blip"/>
                      </t>
                  </t>
              </div>`;
        static components = { Child };
        state = { s: [{ blips: ["a1", "a2"] }, { blips: ["b1"] }] };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<div><div>a1</div><div>a2</div><div>b1</div></div>");
      expect(qweb.compiledTemplates[Parent.template].fn.toString()).toMatchSnapshot();
    });

    test("reconciliation alg works for t-foreach in t-foreach, 2", async () => {
      class Child extends Component {
        static template = xml`<div><t t-esc="props.row + '_' + props.col"/></div>`;
      }

      class Parent extends Component {
        static template = xml`
              <div>
                <p t-foreach="state.rows" t-as="row" t-key="row">
                  <p t-foreach="state.cols" t-as="col" t-key="col">
                    <Child row="row" col="col"/>
                  </p>
                </p>
              </div>`;
        static components = { Child };
        state = useState({ rows: [1, 2], cols: ["a", "b"] });
      }

      const parent = await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe(
        "<div><p><p><div>1_a</div></p><p><div>1_b</div></p></p><p><p><div>2_a</div></p><p><div>2_b</div></p></p></div>"
      );
      parent.state.rows = [2, 1];
      await nextTick();
      expect(fixture.innerHTML).toBe(
        "<div><p><p><div>2_a</div></p><p><div>2_b</div></p></p><p><p><div>1_a</div></p><p><div>1_b</div></p></p></div>"
      );
    });

    test("same t-keys in two different places", async () => {
      class Child extends Component {
        static template = xml`<span><t t-esc="props.blip"/></span>`;
      }

      class Parent extends Component {
        static template = xml`
              <div>
                  <div><Child t-key="1" blip="'1'"/></div>
                  <div><Child t-key="1" blip="'2'"/></div>
              </div>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe(
        "<div><div><span>1</span></div><div><span>2</span></div></div>"
      );
      expect(qweb.compiledTemplates[Parent.template].fn.toString()).toMatchSnapshot();
    });

    test("t-key on a component with t-if, and a sibling component", async () => {
      class Child extends Component {
        static template = xml`<span>child</span>`;
      }

      class Parent extends Component {
        static template = xml`
              <div>
                <Child t-if="false" t-key="'str'"/>
                <Child/>
              </div>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(fixture.innerHTML).toBe("<div><span>child</span></div>");
      expect(qweb.compiledTemplates[Parent.template].fn.toString()).toMatchSnapshot();
    });
  });
});
