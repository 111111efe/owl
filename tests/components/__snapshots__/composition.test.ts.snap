// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component composition various update scenarios reconciliation alg is not confused in some specific situation 1`] = `
"function anonymous(ctx, metadata
) {
    // Template: \`__template__28\`
    let _1 = {};
    let vn3 = this.makeComponent(\`k2\`, metadata, \\"Child\\", ctx, _1);
    let key4 = 4;
    let _5 = {};
    let vn7 = this.makeComponent(\`k6__\${key4}\`, metadata, \\"Child\\", ctx, _5);
    let vn9 = {type: 2, tag: \\"div\\", children: [vn3,vn7], key: \`k8\`};
    return vn9;
}"
`;

exports[`component composition various update scenarios reconciliation alg works for t-foreach in t-foreach 1`] = `
"function anonymous(ctx, metadata
) {
    // Template: \`__template__30\`
    let vn2 = {type: 2, tag: \\"div\\", children: [], key: \`k1\`};
    let _4 = ctx['state'].s;
    let length3 = _4.length;
    ctx = Object.create(ctx);
    for (let i0 = 0; i0 < length3; i0++) {
        ctx.section_first = i0 === 0;
        ctx.section_last = i0 === length3 - 1;
        ctx.section = _4[i0];
        ctx.section_index = i0;
        ctx.section_value = _4[i0];
        let _6 = ctx['section'].blips;
        let length5 = _6.length;
        ctx = Object.create(ctx);
        for (let i1 = 0; i1 < length5; i1++) {
            ctx.blip_first = i1 === 0;
            ctx.blip_last = i1 === length5 - 1;
            ctx.blip = _6[i1];
            ctx.blip_index = i1;
            ctx.blip_value = _6[i1];
            let _7 = {blip:ctx['blip']};
            let vn9 = this.makeComponent(\`k8__\${i0}__\${i1}\`, metadata, \\"Child\\", ctx, _7);
            vn2.children.push(vn9);
        }
        ctx = ctx.__proto__;
    }
    ctx = ctx.__proto__;
    return vn2;
}"
`;

exports[`component composition various update scenarios reconciliation alg works for t-foreach in t-foreach, 2 1`] = `
"function anonymous(ctx, metadata
) {
    // Template: \`__template__32\`
    let vn2 = {type: 2, tag: \\"div\\", children: [], key: \`k1\`};
    let _4 = ctx['state'].rows;
    let length3 = _4.length;
    ctx = Object.create(ctx);
    for (let i0 = 0; i0 < length3; i0++) {
        ctx.row_first = i0 === 0;
        ctx.row_last = i0 === length3 - 1;
        ctx.row = _4[i0];
        ctx.row_index = i0;
        ctx.row_value = _4[i0];
        let key5 = ctx['row'];
        let vn7 = {type: 2, tag: \\"p\\", children: [], key: \`k6__\${key5}\`};
        vn2.children.push(vn7);
        let _9 = ctx['state'].cols;
        let length8 = _9.length;
        ctx = Object.create(ctx);
        for (let i1 = 0; i1 < length8; i1++) {
            ctx.col_first = i1 === 0;
            ctx.col_last = i1 === length8 - 1;
            ctx.col = _9[i1];
            ctx.col_index = i1;
            ctx.col_value = _9[i1];
            let key10 = ctx['col'];
            let _11 = {row:ctx['row'],col:ctx['col']};
            let vn13 = this.makeComponent(\`k12__\${key5}__\${key10}\`, metadata, \\"Child\\", ctx, _11);
            let vn15 = {type: 2, tag: \\"p\\", children: [vn13], key: \`k14__\${key10}\`};
            vn7.children.push(vn15);
        }
        ctx = ctx.__proto__;
    }
    ctx = ctx.__proto__;
    return vn2;
}"
`;

exports[`component composition various update scenarios same t-keys in two different places 1`] = `
"function anonymous(ctx, metadata
) {
    // Template: \`__template__34\`
    let key1 = 1;
    let _2 = {blip:'1'};
    let vn4 = this.makeComponent(\`k3__\${key1}\`, metadata, \\"Child\\", ctx, _2);
    let vn6 = {type: 2, tag: \\"div\\", children: [vn4], key: \`k5\`};
    let key7 = 1;
    let _8 = {blip:'2'};
    let vn10 = this.makeComponent(\`k9__\${key7}\`, metadata, \\"Child\\", ctx, _8);
    let vn12 = {type: 2, tag: \\"div\\", children: [vn10], key: \`k11\`};
    let vn14 = {type: 2, tag: \\"div\\", children: [vn6,vn12], key: \`k13\`};
    return vn14;
}"
`;

exports[`component composition various update scenarios t-key on a component with t-if, and a sibling component 1`] = `
"function anonymous(ctx, metadata
) {
    // Template: \`__template__36\`
    let vn2 = {type: 2, tag: \\"div\\", children: [], key: \`k1\`};
    if (false) {
        let key3 = 'str';
        let _4 = {};
        let vn6 = this.makeComponent(\`k5__\${key3}\`, metadata, \\"Child\\", ctx, _4);
        vn2.children.push(vn6);
    }
    let _7 = {};
    let vn9 = this.makeComponent(\`k8\`, metadata, \\"Child\\", ctx, _7);
    vn2.children.push(vn9);
    return vn2;
}"
`;
